### PR TITLE
Feature #2833 pcp_combine_missing

### LINKED ISSUE #2883: Enhance PCP-Combine to allow missing data
## Describe the Enhancement ##
- Modify PCP-Combine to allow missing input files with a configurable threshold (vld_thresh) 
- If a file is missing but its path includes the `MISSING` keyword, do not print a warning message.
- If a file is missing without the `MISSING` keyword, do print a warning message.
- Note that Pcp-Combine already includes support for file lists. So the Pcp-Combine METplus wrapper could be updated to provide input through file lists.

Data to be used for testing can be found at: `seneca:/d1/projects/METplus/METplus_Data/development/pcp_combine_missing`

Ensure that missing files are handled properly when they are found at the beginning, end, or middle of the list of input files.

### Time Estimate ###
2 days

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
None.

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2799991

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Review default **alert** labels
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Milestone and Projects ###
- [x] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [x] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] dtcenter/METplus#2578 updates the PcpCombine wrapper accordingly

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
Adds new `-input_thresh` command line option for pcp_combine which is supported for the `-add`, `-derive`, and `-sum` options. It's NOT supported for `-subtract` for which exactly two valid input files are required.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Added 1 new `pcp_combine_sum_GRIB1_MISSING` test in `unit_pcp_combine.xml` to demonstrate running `-sum` with one of the 4 expected inputs missing but `-input_thresh 0.75` set to allow output to be written.

Modified 1 existing `pcp_combine_derive_MULTIPLE_FIELDS` test in `unit_pcp_combine.xml` to demonstrate running `-derive` with missing inputs files at the beginning, middle, and end of the input file list. The first 2 do not trigger warnings because the MISSING keyword is used. The 3rd does, because MISSING is not present in the path.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Please review the code changes, documentation updates, and testing changes.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

- [x] Will this PR result in changes to the MET test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
One new output file named `pcp_combine/nam_2012040900_F087_APCP12.nc` is created.

- [x] Will this PR result in changes to existing METplus Use Cases? **[No]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.

- [x] Do these changes introduce new SonarQube findings? **[Yes or No]**</br>
If **yes**, please describe:
I don't know. If the GHA run for this PR triggers them, I'll go fix them.

- [x] Please complete this pull request review by **[Tues 5/14/24]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: docs/Users_Guide/reformat_grid.rst [hunk 1]
```diff
@@ -79,13 +80,15 @@ Optional Arguments for pcp_combine
 
 4. The **-name list** option is a comma-separated list of output variable names which override the default choices. If specified, the number of names must match the number of variables to be written to the output file.
 
-5. The **-vld_thresh n** option overrides the default required ratio of valid data for at each grid point for an output value to be written. The default is 1.0.
+5. The **-input_thresh n** option overrides the default required ratio of valid input files. This option does not apply to the -sub command where exactly two valid inputs are required. The default is 1.0.
```
Comment (georgemccabe) [inline docs/Users_Guide/reformat_grid.rst]:
```suggestion
5. The **-input_thresh n** option overrides the default required ratio of valid input files. This option does not apply to the -subtract command where exactly two valid inputs are required. The default is 1.0.
```

File: src/tools/core/pcp_combine/pcp_combine.cc [hunk 2]
```diff
@@ -1546,6 +1626,9 @@ void usage() {
         << "variable names to be written to the \"out_file\" "
         << "(optional).\n"
 
+        << "\t\t\"-input_thresh\" overrides the default required ratio "
+        << "of valid input files (" << input_thresh << ") (optional).\n"
+
         << "\t\t\"-vld_thresh\" overrides the default required ratio "
         << "of valid data (" << vld_thresh << ") (optional).\n"
```
Comment (georgemccabe) [inline src/tools/core/pcp_combine/pcp_combine.cc]: Consider refining this to make it more clear the distinction between input and vld thresh
```suggestion
        << "of valid data at a grid point (" << vld_thresh << ") (optional).\n"
```

**Review by georgemccabe — CHANGES_REQUESTED** [review]
These changes look good. I made a few minor suggestions for the documentation and usage statement.

File: src/tools/core/pcp_combine/pcp_combine.cc (hunk 2 above)
Comment (JohnHalleyGotway) [reply src/tools/core/pcp_combine/pcp_combine.cc]: Updated via a change to my branch. Thanks.

**Review by georgemccabe — CHANGES_REQUESTED** [review]
It looks like the `pcp_combine_derive_VLD_THRESH` unit test is failing because the previous test was modified to change the list of input files to include missing paths and this test uses the same file list file. To fix this we could either:
1) Create another file list text file that matches the previous list to use in this test
OR
2) Add the `-input_thresh` argument to this test

It looks like the SonarQube test also failed.

**Comment by JohnHalleyGotway** [comment]
Converted PR to draft while I fix the tests and SonarQube findings.

**Comment by JohnHalleyGotway** [comment]
@georgemccabe I just reopened this PR after fixing the testing and SonarQube issues.

Note that [14 SonarQube code smells](https://needham.rap.ucar.edu/dashboard?id=MET&branch=refs%2Fpull%2F2886%2Fmerge) remain for this PR, but these are ones I do not plan to fix. They're not easy, and I confirmed with Julie and Michelle that these don't have a high enough priority level to be worked on right now.

I do expect differences in the output to be flagged, but only based on the addition of 1 new file.

**Review by georgemccabe — CHANGES_REQUESTED** [review]
It looks like another unit test is failing: python_pcp_combine_SEMILATLON_MERIDIONAL_MEAN

**Comment by JohnHalleyGotway** [comment]
Phew, OK, this[ testing workflow run](https://github.com/dtcenter/MET/actions/runs/9087391809/job/24975748978) now completes all the tests and only flags one new output file, as expected:
```
ERROR: folder /data/output/met_test_truth missing 1 files
    pcp_combine/nam_2012040900_F087_APCP12.nc 
```
The updated logic in `parse_file_list.cc` is working as expected.
This PR is ready to be re-reviewed.

**Review by georgemccabe — APPROVED** [review]
Looks good, thanks for making those fixes.

**Pull request merged**